pkg/codegen/dotnet/gen_program.go

```diff
@@ -0,0 +1,428 @@
+// and the count variable name, if any.
+func (g *generator) makeResourceName(baseName, count string) string {
+	if count == "" {
+		return fmt.Sprintf(`"%s"`, baseName)
+	}
+	return fmt.Sprintf("$\"%s-{%s}\"", baseName, count)
+}
+
+// genResource handles the generation of instantiations of non-builtin resources.
+func (g *generator) genResource(w io.Writer, r *hcl2.Resource) {
+	qualifiedMemberName := g.resourceTypeName(r)
+
+	// Add conversions to input properties
+	for _, input := range r.Inputs {
+		destType, diagnostics := r.InputType.Traverse(hcl.TraverseAttr{Name: input.Name})
+		g.diagnostics = append(g.diagnostics, diagnostics...)
+		input.Value = g.rewriteExpression(input.Value, destType.(model.Type))
+	}
+
+	optionsBag := ""
+
+	name := r.Name()
+
+	g.genTrivia(w, r.Definition.Tokens.GetType(""))
+	for _, l := range r.Definition.Tokens.GetLabels(nil) {
+		g.genTrivia(w, l)
+	}
+	g.genTrivia(w, r.Definition.Tokens.GetOpenBrace())
+
+	instantiate := func(resName string) {
+		g.Fgenf(w, "new %s(%s, new %[1]sArgs\n", qualifiedMemberName, resName)
+		g.Fgenf(w, "%s{\n", g.Indent)
+		g.Indented(func() {
+			for _, attr := range r.Inputs {
+				g.Fgenf(w, "%s%s =", g.Indent, propertyName(attr.Name))
+				g.Fgenf(w, " %.v,\n", attr.Value)
+			}
+		})
+		g.Fgenf(w, "%s}%s)", g.Indent, optionsBag)
+	}
+
+	if r.Options != nil && r.Options.Range != nil {
+		g.Fgenf(w, "%svar %s = new List<%s>();\n", g.Indent, name, qualifiedMemberName)
+
+		rangeExpr := &model.FunctionCallExpression{
+			Name: "entries",
+			Args: []model.Expression{r.Options.Range},
+		}
+		g.Fgenf(w, "%sforeach (var range in %.v)\n", g.Indent, rangeExpr)
+		g.Fgenf(w, "%s{\n", g.Indent)
+
+		resName := g.makeResourceName(name, "range.Key")
+		g.Indented(func() {
+			g.Fgenf(w, "%s%s.Add(", g.Indent, name)
+			instantiate(resName)
+			g.Fgenf(w, ");\n")
+		})
+		g.Fgenf(w, "%s}\n", g.Indent)
+	} else {
+		g.Fgenf(w, "%svar %s = ", g.Indent, name)
+		instantiate(g.makeResourceName(name, ""))
+		g.Fgenf(w, ";\n")
+	}
+
+	g.genTrivia(w, r.Definition.Tokens.GetCloseBrace())
+}
+
+func (g *generator) genConfigVariable(w io.Writer, v *hcl2.ConfigVariable) {
+	g.genNYI(w, "// TODO config")
+}
+
+func (g *generator) genLocalVariable(w io.Writer, v *hcl2.LocalVariable) {
+	// TODO(pdg): trivia
+	expr := g.rewriteExpression(v.Definition.Value, v.Type())
+	g.Fgenf(w, "%svar %s = %.3v;\n", g.Indent, validIdentifier(v.Name()), expr)
+}
+
+func (g *generator) genOutputAssignment(w io.Writer, v *hcl2.OutputVariable) {
+	if g.asyncInit {
+		g.Fgenf(w, "%svar %s", g.Indent, validIdentifier(v.Name()))
+	} else {
+		g.Fgenf(w, "%sthis.%s", g.Indent, propertyName(v.Name()))
+	}
+	g.Fgenf(w, " = %.3v;\n", g.rewriteExpression(v.Value, v.Type()))
+}
+
+func (g *generator) genOutputProperty(w io.Writer, v *hcl2.OutputVariable) {
+	// TODO(pdg): trivia
+	g.Fgenf(w, "%s[Output(\"%s\")]\n", g.Indent, v.Name())
+	g.Fgenf(w, "%spublic Output<string> %s { get; set; }\n", g.Indent, propertyName(v.Name()))
```

Comment:
The element type of the `Output` should be derived from the type of its value instead of being hardcoded to `string`, right?

Reply:
Good catch!

Reply:
I _think_ we are missing required type annotations for this. I added a TODO and will merge as-is to handle this in a follow-up.